File: leetcode/adjacent_duplicates.cpp
### cpp equivalent of https://leetcode.com/problems/remove-all-adjacent-duplicates-in-string/description/
#include<iostream>
#include<stack>

using namespace std;
class Solution {
public:
    string removeDuplicates(string s) {
        stack<char> stack;
        std::string result = "";
        for(int i=0; i< s.size(); i++){
            if(stack.empty() == 0){
                if(stack.top() != s[i]){
                    stack.push(s[i]);
                }else{
                    stack.pop();
                }
            }else{
                stack.push(s[i]);
            }
        }
        while(!stack.empty()){
            char c = stack.top();
            stack.pop();
            result += c;
        }
        reverse(result.begin(), result.end());
        return result;
    }
};

int main(){
    Solution s = Solution();
    std::string result = s.removeDuplicates("abbaca");
    std::cout<<result<<std::endl;
    return 0;
}

File: leetcode/adjacent_duplicates.py
# Problem: https://leetcode.com/problems/remove-all-adjacent-duplicates-in-string/

class Solution:
    def removeDuplicates(self, s: str) -> str:
        stack = []
        for letter in s:
            if(len(stack) != 0):
                top = stack[-1]
                if letter == top:
                    stack.pop()
                else:
                    stack.append(letter)
            else:
                stack.append(letter)
        result = "".join(stack);
        return result;

if __name__ == '__main__':
    s = Solution();
    result = s.removeDuplicates('azxxzy');
    print(result);
    result = s.removeDuplicates('aabccba');
    print(result);
    result = s.removeDuplicates('aacbcca');
    print(result);

### output
# mahesh@Maheshs-MacBook-Air-M1 leetcode % python3 adjacent_duplicates.py
# ay
# a
# cba
# mahesh@Maheshs-MacBook-Air-M1 leetcode % 